docs/blueprints/EP036.rst

```diff
@@ -0,0 +1,164 @@
+    |                  |         |                  |         |                  |         |                  |         |                  |         |                  |
+    |   goto_table: 1  |         |   goto_table: 2  |         |   goto_table: 3  |         |   goto_table: 5  |         |                  |         |                  |
+    +------------------+         +------------------+         +------------------+         +------------------+         +------------------+         +------------------+
+
+New collection, **tables**
+==========================
+
+Initially proposed here is a datastructured representing a linked list (without cycles)
+
+  .. code:: javascript
+
+    "pipeline": {
+      0: {
+        "table_miss_flow": {
+          "priority": 123, "match": {},
+          "instructions": [{"instruction_type": "goto_table", "table_id": 1}]
+        },
+        "napps_table_groups": {"coloring": ["base"],"of_lldp": ["base"]}
+      },
+      1: {
+        "table_miss_flow": {
+          "priority": 123, "match": {},
+          "instructions": [{"instruction_type": "goto_table","table_id": 2}]
+        },
+        "napps_table_groups": {"flow_manager": []}
+      },
+      2: {
+        "table_miss_flow": {
+          "priority": 123, "match": {},
+          "instructions": [{"instruction_type": "goto_table","table_id": 3}]
+        },
+        "napps_table_groups": {"mef_eline": ["evpl"]}
+      },
+      3: {
+        "table_miss_flow": {
+          "priority": 123, "match": {},
+          "instructions": [{"instruction_type": "goto_table","table_id": 5}]
+        },
+        "napps_table_groups": {"mef_eline": ["epl"]}
+      }
+      4: {
+        "table_miss_flow": {},
+        "napps_table_groups": {"INT": []}
+      }
+      5: {
+        "table_miss_flow": {},
+        "napps_table_groups": {}
+      }
+    }
+
+Events
+======
+
+Subscribed
+-----------
+  - **kytos/flow_manager.flow.added**
+  - **kytos/of_core.handshake.completed**
+
+Published
+---------
+
+kytos/multi_table.deploy_table
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This event should sent before any NApp sends its first flow to be published.
+
+.. code:: javascript
+
+  "content": {
+    "mef_eline": {"epl": 3, "evpl": 2},
+    "of_lldp": {"base": 0},
+    "coloring": {"base": 0}
+  }
+
+Deploying pipeline
+==================
+
+Im simple terms, deploying a pipeline will mean to modify every flow so it complies with the pipeline configuration. The process to achieve a successful deployment is as follows:
+
+  1. **multi_table** publishes `kytos/multi_table.deploy_table` event with pipeline content.
+  2. **multi_table** requests all flows from **flow_manager** API.
+  3. **multi_table** duplicates every flow with a different table (to achieve this, this NApp needs to be aware of every other NApp cookie to identify ownership).
+  4. The previous flows are deleted from the database

```

Comment:
`multi_table` table miss flows we'll always need to delete/reinstall though to be safe as we've brainstormed.

Reply:
> For NApps flow deletions, is it only going to delete flows from table_groups that changed their tables? If that's the case, we might not even need to rely on consistency check to reinstall things.

Yeah, I am going to specify that better now. If we do not want to rely on consistency check we could mark flows to be ignored, add flows with different table, delete flows and add flows, Is this the correct order?. I am still not sure what triggers the consistency check but would be better to avoid the reinstalling again the flow that `multi_table` recently added.

Reply:
The least disrupting order would generally be:

- Reinstall flows that were moved to a different table. 
- Delete old flows entries that were moved (on the prior step). 

> we could mark flows to be ignored

If you mean ignore flows that haven't changed then yes. But if you're referring to the actual ignore mechanism of cookies and table ids that `flow_manager` exposes too, that's only for flow entries that flow_manager shouldn't even be managing and only literally just ignoring when receiving the flow stats. 

> triggers the consistency check. Is there a way for a NApp to stop consistency check?

In summary, a switch reconnection or after flow stats are received, if recent flows were installed or deleted it'll skip iterations though to minimize false positive during convergence. Consistency check has a settings value that could be disabled in memory too, although by design the consistency check should be smart enough not to prematurely kicks in so not requiring to be disabled or changed during runtime.

Reply:
Since we're dealing with a concurrency and distributed system problem here, we might also need to define another events like `kytos/<napp_name>.deploy_table.error`, and `kytos/<napp_name>.deploy_table.ok` (or something along these lines) just so `multi_table` can subscribe to confirm that `kytos/multi_table.deploy_table` either errored out or succeeded . Consequently, `multi_table` shouldn't start to request the flows and proceed with the installation/removals for a given pipeline id until all NApps have confirmed.